File: with_rust/src/day4.rs
#[rustfmt::skip]
const DIRECTIONS: [(i32, i32); 8] = [
    (-1, -1), (-1, 0), (-1, 1),
     (0, -1),          (0, 1),
     (1, -1),  (1, 0), (1, 1)
];

pub fn part1(content: Vec<String>) -> i32 {
    let h = content.len() as i32;
    let w = content[0].len() as i32;

    let is_inbound = |x: i32, y: i32| x >= 0 && x < h && y >= 0 && y < w;

    let check_xmas = |x: i32, y: i32, h_dir: i32, v_dir: i32| -> bool {
        (0..4)
            .map(|i| (x + i * h_dir, y + i * v_dir))
            .collect::<Vec<(i32, i32)>>()
            .iter()
            .enumerate()
            .all(|(i, &(x, y))| {
                is_inbound(x, y)
                    && content[x as usize].chars().nth(y as usize) == "XMAS".chars().nth(i)
            })
    };

    content
        .iter()
        .enumerate()
        .flat_map(|(x, row)| {
            row.chars().enumerate().filter_map(move |(y, c)| {
                if c == 'X' {
                    Some((x as i32, y as i32))
                } else {
                    None
                }
            })
        })
        .flat_map(|(x, y)| {
            DIRECTIONS
                .iter()
                .filter(move |&&(h_dir, v_dir)| check_xmas(x, y, h_dir, v_dir))
        })
        .count() as i32
}

const PATTERNS: [[char; 4]; 4] = [
    ['M', 'M', 'S', 'S'],
    ['S', 'S', 'M', 'M'],
    ['M', 'S', 'M', 'S'],
    ['S', 'M', 'S', 'M'],
];

pub fn part2(content: Vec<String>) -> i32 {
    let h = content.len();
    let w = content[0].len();

    let get_symbols_around = |x: usize, y: usize| -> Option<[char; 4]> {
        if x == 0 || y == 0 || x >= h - 1 || y >= w - 1 {
            return None;
        }

        Some([
            content[x - 1].chars().nth(y - 1).expect("Invalid top-left"),
            content[x - 1]
                .chars()
                .nth(y + 1)
                .expect("Invalid top-right"),
            content[x + 1]
                .chars()
                .nth(y - 1)
                .expect("Invalid bottom-left"),
            content[x + 1]
                .chars()
                .nth(y + 1)
                .expect("Invalid bottom-right"),
        ])
    };

    let check_patterns = |symbols_around: [char; 4]| -> bool {
        PATTERNS.iter().any(|pattern| pattern == &symbols_around)
    };

    content
        .iter()
        .enumerate()
        .skip(1)
        .take(h - 2)
        .flat_map(|(x, row)| {
            row.chars()
                .enumerate()
                .skip(1)
                .take(w - 2)
                .filter_map(move |(y, c)| {
                    if c == 'A' {
                        get_symbols_around(x, y)
                            .map(|symbols_around| check_patterns(symbols_around))
                    } else {
                        None
                    }
                })
        })
        .filter(|&valid| valid)
        .count() as i32
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::reader;

    #[test]
    fn part1solution() {
        assert_eq!(part1(reader::read_input("day4sample")), 18);
    }

    #[test]
    fn part2solution() {
        assert_eq!(part2(reader::read_input("day4sample")), 9);
    }
}

File: 2024/with_python/src/day4.py
def part1(content: list[str]) -> int:
    # fmt: off
    DIRECTIONS = [
        (-1, -1), (-1, +0), (-1, +1),
        (+0, -1),           (+0, +1),
        (+1, -1), (+1, +0), (+1, +1),
    ]
    # fmt: on

    h = len(content)
    w = len(content[0])

    def is_in_bounds(x: int, y: int) -> bool:
        return 0 <= x < w and 0 <= y < h

    def check_xmas(x: int, y: int, h_dir: int, v_dir: int) -> bool:
        positions = [(x + i * h_dir, y + i * v_dir) for i in range(4)]

        if not all(is_in_bounds(x, y) for x, y in positions):
            return False
        target = "XMAS"
        return all(content[px][py] == target[i] for i, (px, py) in enumerate(positions))

    return sum(
        check_xmas(x, y, h_dir, v_dir)
        for x in range(h)
        for y in range(w)
        for h_dir, v_dir in DIRECTIONS
        if content[x][y] == "X"
    )


def part2(content: list[str]) -> int:
    def check_patterns(x: int, y: int) -> bool:
        if content[x][y] != "A":
            return False

        PATTERNS = [
            {"tl": "M", "bl": "M", "tr": "S", "br": "S"},
            {"tl": "S", "bl": "S", "tr": "M", "br": "M"},
            {"tl": "M", "bl": "S", "tr": "M", "br": "S"},
            {"tl": "S", "bl": "M", "tr": "S", "br": "M"},
        ]

        symbols_around_A = {
            "tl": content[x - 1][y - 1],
            "tr": content[x - 1][y + 1],
            "bl": content[x + 1][y - 1],
            "br": content[x + 1][y + 1],
        }

        return symbols_around_A in PATTERNS

    h = len(content)
    w = len(content[0])

    return sum(check_patterns(x, y) for x in range(1, h - 1) for y in range(1, w - 1))
